portfolio/src/main/webapp/comments.js

```diff
@@ -0,0 +1,129 @@
+  // create a hash table of type "comment id" -> "comment" to build a comment tree
+  const commentsMap = {};
+  for (const i in commentsJson) {
+    const comment = commentsJson[i];
+    commentsMap[comment.key] = comment;
+  }
+  if (commentsJson.length) {
+    // if there are some comments, remove label 'No comments yet'
+    document.getElementById('no-comments').remove();
+  }
+  // build a tree, starting DFS from the vertex (comments) that have no incoming vertexes (are not replies)
+  for (const i in commentsMap) {
+    const comment = commentsMap[i];
+    if (!comment.replyTo) {
+      buildCommentsTree(commentsMap, comment, 0);
+    }
+  }
+}
+
+function buildCommentsTree(commentsMap, comment, depth) {
+  // add current comment to dom
+  addCommentToDOM(comment, depth)
+  // iterate over all replies in the map, add them as the children of the current comment
+  for (const i in comment.replies) {
+    const replyId = comment.replies[i];
+    const reply = commentsMap[replyId]
+    buildCommentsTree(commentsMap, reply, Number(depth + 1));
+  }
+}
+
+function addCommentToDOM(comment, depth) {
+  // creates a div with two paragraph, one for username and one for comment text
+  const commentDOMElement = document.createElement('div');
+  commentDOMElement.classList.add('comment')
+  commentDOMElement.style = 'margin-left: ' + Math.sqrt(Number(depth * 100)) + 'px';
+  commentDOMElement.id = comment.key;
+  const commentText = document.createElement('p');
+  commentText.innerText = comment.text;
+  const showReplyFormButton = document.createElement('button');
+  showReplyFormButton.innerText = 'Reply';
+  showReplyFormButton.id = comment.key + '_button'; // this button must have unique id as I will change it's text
+                                                    // depending on form state
+  showReplyFormButton.onclick = function() {
+    toggleReplyField(comment.key);
+  };
+  const replyForm = getCommentReplyForm(comment);
+  const commentHeader = getCommentHeader(comment);
+  // append everything to the comment div
+  commentDOMElement.appendChild(commentHeader);
+  commentDOMElement.appendChild(commentText);
+  commentDOMElement.appendChild(replyForm);
+  commentDOMElement.appendChild(showReplyFormButton);
+  // attach created comment to the comments div
+  document.getElementById('comments').appendChild(commentDOMElement);
+}
+
+function getCommentHeader(comment) {
+  const commentAuthorUsername = document.createElement('h5');
+  commentAuthorUsername.innerText = comment.username;
+  const commentDate = document.createElement('p');
+  commentDate.innerText = humanReadableDateFromTimestamp(comment.timestamp);
+  // header for both username and date
+  const commentHeader = document.createElement('div');
+  commentHeader.classList.add('comment-header');
+  commentHeader.appendChild(commentAuthorUsername);
+  commentHeader.appendChild(commentDate);
+  return commentHeader;
+}
+
+function getCommentReplyForm(comment) {
```

Comment:
Same comment here too. A good rule of thumb is to have  not more than 15-20 different computation statements in a method description. The smaller they are, they easier they are to read and review. Also they are easy to test.

Reply:
What would you suggest to make this metod shorter? It is a one logical block - it creates a reply form. Do you suggest to move creation of different parts of this form into separate methods? Like for example, write method that creates an input field for the form, or method creates a button for this form?

Reply:
> write method that creates an input field for the form

This sounds like a nice way to shorten this method.
const formReplyUsername = whateverMethodName('text', 'username', 'Your username');
....
infact if you do it this way, I think you can even remove the need for constants and directly do replyForm.appendChild(whateverMethodName(...)); multiple times